Improve TimeSeriesBase docstring

dstansby opened the pull request.

pep8speaks commented:
Hello @dstansby! Thanks for submitting the PR.

 - In the file [`sunpy/timeseries/timeseriesbase.py`](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py), following are the PEP8 issues :

> [Line 134:9](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L134): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 135:9](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L135): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 180:13](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L180): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 212:13](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L212): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 213:13](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L213): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 242:101](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L242): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (138 > 100 characters)
> [Line 273:16](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L273): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 277:16](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L277): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 314:101](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L314): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 320:101](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L320): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 438:101](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L438): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 100 characters)
> [Line 511:33](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L511): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 608:101](https://github.com/sunpy/sunpy/blob/49744ad288286433f3aced8af00959a52aa6ed77/sunpy/timeseries/timeseriesbase.py#L608): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)

dstansby commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -68,25 +68,23 @@ class GenericTimeSeries:
-
-    Parameters
-    ----------------
-    filename: `str` or File
```
I have no idea what all these parameters are doing here in the docstring, but they don't seem to be used anywhere in the actual code...

dstansby commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -140,23 +138,27 @@ def __init__(self, data, meta=None, units=None, **kwargs):
     def source(self):
-        """Returns a string/object used to specify the source class of the TimeSeries."""
+        """
+        A string/object used to specify the source class of the TimeSeries.
```
Properties show up as attributes in the docs, so don't need a "Returns..." at the beginning of the description.

Cadair commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -68,25 +68,23 @@ class GenericTimeSeries:
+    meta : `~sunpy.timeseries.metadata.TimeSeriesMetaData`, optional
+        The metadata giving details about the time series data/instrument.
+    **kwargs : dict
+        Keyword arguments are not used.
```
You can remove this from here, they don't need documenting if they don't do anything.

IIRC, the reason `**kwargs` is here is because sometimes the factory passes extra things through, which maybe it shouldn't but is not a trivial fix.

dstansby commented:
Removed the **kwargs bit, and also added property functions for data/meta, as this means they show up better in the docs (and I don't know of any reason from a code perspective not to?)

dpshelio commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -138,25 +126,43 @@ def __init__(self, data, meta=None, units=None, **kwargs):
+        """
+        The original pandas DataFrame.
+        """
+        return self.data
```
I may be getting something wrong, but `data` and `meta` aren't already public attributes as they were? 
In any case, why if I do the following doesn't work?
```python
class MyClass(object):
    def __init__(self, data, meta):
        self.data = data
        self.meta = meta 
    @property
    def _data(self):
        return self.data
```
Should the `self.data` be renamed as `self._data` (making it private)? Wouldn't then you also need some setter? Is that to do some data sanity?

dstansby replied:
Yeah, I realised that what I did doesn't actually work, I'll revert it...

nabobalis commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -68,25 +68,21 @@ class GenericTimeSeries:
         The metadata giving details about the time series data/instrument.
-    data : `~pandas.DataFrame`
-        An pandas DataFrame prepresenting one or more fields as a function of time.
-
```
Do we need both Attributes and Parameters?

dstansby replied:
Attributes are class attributes, and parameters are things passed to `__init__`, so they're different?

nabobalis replied:
Just that they seemed the same, can't we merge the two into one?

Cadair replied:
yeah they need to be in both places.

wafels commented:
Isn't 'data' is a parameter, and 'meta' a keyword in the __init__ ? Also,
there is no description of what the units keyword does, and why is there a
**kwargs in the __init__?

On Wed, Jan 24, 2018 at 7:18 AM, Nabil Freij <notifications@github.com>
wrote:

> *@nabobalis* commented on this pull request.
> ------------------------------
>
> In sunpy/timeseries/timeseriesbase.py
> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163531135>:
>
> >      meta : `~sunpy.timeseries.metadata.TimeSeriesMetaData`
>          The metadata giving details about the time series data/instrument.
> -    data : `~pandas.DataFrame`
> -        An pandas DataFrame prepresenting one or more fields as a function of time.
> -
>
> Just that they seemed the same, can't we merge the two into one?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163531135>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA8CFzMborWyR7Hocdu5u133El3ddXArks5tNx96gaJpZM4Ravad>
> .
>

dstansby commented:
I thought all the inputs to `__init__` are parameters by definition?

For the other two questions, I don't know what units does so I haven't added a description, and I have no idea why there's a **kwargs. Feel free to push to my branch if these need changing.

Cadair commented on sunpy/timeseries/timeseriesbase.py:
```diff
@@ -68,25 +68,21 @@ class GenericTimeSeries:
+        of time.
+    meta : `~sunpy.timeseries.metadata.TimeSeriesMetaData`, optional
+        The metadata giving details about the time series data/instrument.
+
```
units should be added in both parameters and attributes. Units are a dictionary of column name : unit .

nabobalis replied:
This is the final issue before merging?

wafels commented:
At the developers meeting today, we were told that **kwargs is fine as it
is.

On Wed, Jan 24, 2018 at 12:44 PM, Stuart Mumford <notifications@github.com>
wrote:

> *@Cadair* commented on this pull request.
> ------------------------------
>
> In sunpy/timeseries/timeseriesbase.py
> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163624290>:
>
> >      meta : `~sunpy.timeseries.metadata.TimeSeriesMetaData`
>          The metadata giving details about the time series data/instrument.
> -    data : `~pandas.DataFrame`
> -        An pandas DataFrame prepresenting one or more fields as a function of time.
> -
>
> yeah they need to be in both places.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163624290>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA8CF85U2_ucf7YOONyODNARboZYlSnTks5tN2wKgaJpZM4Ravad>
> .
>

wafels commented:
It has to be there, and it gets passed along to the TimeSeries factory, if
I understood correctly.  So I think the only thing missing is a description
of units keyword.

On Wed, Jan 24, 2018 at 12:46 PM, Jack Ireland <jackireland@gmail.com>
wrote:

> At the developers meeting today, we were told that **kwargs is fine as it
> is.
>
> On Wed, Jan 24, 2018 at 12:44 PM, Stuart Mumford <notifications@github.com
> > wrote:
>
>> *@Cadair* commented on this pull request.
>> ------------------------------
>>
>> In sunpy/timeseries/timeseriesbase.py
>> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163624290>:
>>
>> >      meta : `~sunpy.timeseries.metadata.TimeSeriesMetaData`
>>          The metadata giving details about the time series data/instrument.
>> -    data : `~pandas.DataFrame`
>> -        An pandas DataFrame prepresenting one or more fields as a function of time.
>> -
>>
>> yeah they need to be in both places.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/sunpy/sunpy/pull/2399#discussion_r163624290>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AA8CF85U2_ucf7YOONyODNARboZYlSnTks5tN2wKgaJpZM4Ravad>
>> .
>>
>
>

nabobalis commented:
Looks good to me.
